src/Stratis.Bitcoin.Features.PoA/PoAFeature.cs

```diff
@@ -1,55 +1,89 @@
 using Stratis.Bitcoin.Interfaces;
+using Stratis.Bitcoin.P2P.Peer;
 using Stratis.Bitcoin.P2P.Protocol.Payloads;
+using Stratis.Bitcoin.Utilities;
 
 namespace Stratis.Bitcoin.Features.PoA
 {
-    /// <summary>
-    /// A class providing extension methods for <see cref="IFullNodeBuilder"/>.
-    /// </summary>
-    public static class FullNodeBuilderConsensusExtension
+    public class PoAFeature : FullNodeFeature
     {
-        public class PoAFeature : FullNodeFeature
+        /// <summary>Manager of node's network connections.</summary>
+        private readonly IConnectionManager connectionManager;
+
+        /// <summary>Thread safe chain of block headers from genesis.</summary>
+        private readonly ConcurrentChain chain;
+
+        private readonly FederationManager federationManager;
+
+        /// <summary>Provider of IBD state.</summary>
+        private readonly IInitialBlockDownloadState initialBlockDownloadState;
+
+        private readonly IConsensusManager consensusManager;
+
+        /// <summary>A handler that can manage the lifetime of network peers.</summary>
+        private readonly IPeerBanning peerBanning;
+
+        /// <summary>Factory for creating loggers.</summary>
+        private readonly ILoggerFactory loggerFactory;
+
+        public PoAFeature(FederationManager federationManager, PayloadProvider payloadProvider, IConnectionManager connectionManager, ConcurrentChain chain,
+            IInitialBlockDownloadState initialBlockDownloadState, IConsensusManager consensusManager, IPeerBanning peerBanning, ILoggerFactory loggerFactory)
         {
-            private readonly FederationManager federationManager;
+            this.federationManager = federationManager;
+            this.connectionManager = connectionManager;
+            this.chain = chain;
+            this.initialBlockDownloadState = initialBlockDownloadState;
+            this.consensusManager = consensusManager;
+            this.peerBanning = peerBanning;
+            this.loggerFactory = loggerFactory;
+
+            payloadProvider.DiscoverPayloads(this.GetType().Assembly);
+        }
 
-            public PoAFeature(FederationManager federationManager, PayloadProvider payloadProvider)
-            {
-                this.federationManager = federationManager;
+        /// <inheritdoc />
+        public override Task InitializeAsync()
+        {
+            NetworkPeerConnectionParameters connectionParameters = this.connectionManager.Parameters;
+            bool oldCMBRemoved = connectionParameters.TemplateBehaviors.Remove(connectionParameters.TemplateBehaviors.Single(x => x is ConsensusManagerBehavior));
+            Guard.Assert(oldCMBRemoved);
 
-                payloadProvider.DiscoverPayloads(this.GetType().Assembly);
-            }
+            connectionParameters.TemplateBehaviors.Add(new PoAConsensusManagerBehavior(this.chain, this.initialBlockDownloadState, this.consensusManager, this.peerBanning, this.loggerFactory));
```

Comment:
all the items in `PoAConsensusManagerBehavior` are injectable, makes me wonder if DI is best here, this will solve the problem by injecting behaviours at node creation.

Do it only if its easy.

Reply:
MithrilMan wants to put them on DI some time in the future. But before that I'm doing it the old way 🙂

Reply:
yeah he already did in the dns task so he will get to this eventually np